alf/experience_replayers/replay_buffer.py

```diff
@@ -55,39 +156,82 @@ def get_batch(self, batch_size, batch_length):
+                env_ids, pos = self._uniform_sample(batch_size, batch_length)
 
-            r = torch.rand(*env_ids.shape)
+            info = BatchInfo(env_ids=env_ids, positions=pos)
 
-            num_positions = self._current_size - batch_length + 1
-            num_positions = num_positions[env_ids]
-            pos = (r * num_positions).to(torch.int64)
-            pos += (self._current_pos - self._current_size)[env_ids]
             pos = pos.reshape(-1, 1)  # [B, 1]
             pos = pos + torch.arange(batch_length).unsqueeze(0)  # [B, T]
             pos = pos % self._max_length
-            env_ids = env_ids.reshape(-1, 1).repeat(1, batch_length)  # [B, T]
+            env_ids = env_ids.reshape(-1, 1).expand(batch_size,
+                                                    batch_length)  # [B, T]
             result = alf.nest.map_structure(lambda b: b[(env_ids, pos)],
                                             self._buffer)
-        return convert_device(result)
+
+            if self._prioritized_sampling:
+                indices = self._env_id_pos_to_index(info.env_ids,
+                                                    info.positions)
+                info = info._replace(
+                    importance_weights=self._sum_tree[indices])
+
+        return convert_device(result), convert_device(info)
+
+    def _uniform_sample(self, batch_size, batch_length):
+        min_size = self._current_size.min()
+        assert min_size >= batch_length, (
+            "Not all environments have enough data. The smallest data "
+            "size is: %s Try storing more data before calling get_batch" %
+            min_size)
+
+        batch_size_per_env = batch_size // self._num_envs
+        remaining = batch_size % self._num_envs
+        if batch_size_per_env > 0:
+            env_ids = torch.arange(self._num_envs).repeat(batch_size_per_env)
+        else:
+            env_ids = torch.zeros(0, dtype=torch.int64)
+        if remaining > 0:
+            eids = torch.randperm(self._num_envs)[:remaining]
+            env_ids = torch.cat([env_ids, eids], dim=0)
+
+        r = torch.rand(*env_ids.shape)
+
+        num_positions = self._current_size - batch_length + 1
+        num_positions = num_positions[env_ids]
+        pos = (r * num_positions).to(torch.int64)
+        pos += (self._current_pos - self._current_size)[env_ids]
+        pos = pos % self._max_length
+
+        return env_ids, pos
+
+    def _prioritized_sample(self, batch_size, batch_length):
+        if batch_length != self._mini_batch_length:
+            if self._mini_batch_length > 1:
+                warning_once(
+                    "It is not advisable to use different batch_length"
+                    "for different calls to get_batch(). Previous batch_length=%d "
+                    "new batch_length=%d" % (self._mini_batch_length,
+                                             batch_length))
+            self._change_mini_batch_length(batch_length)
```

Comment:
If I've added many batches with mini_batch_len=1 but sample with mini_batch_len>1, will there be any problem?

Reply:
We currently only support adding batch_length==1.  It is sampling that this comment is talking about.

Reply:
But there is a function that changes mini_batch_length in this PR

Reply:
OK I see.